src/libsyntax/ext/expand.rs

```diff
@@ -836,9 +935,13 @@ impl<'a, 'b> Folder for InvocationCollector<'a, 'b> {
 
         let (item, attr) = self.classify_item(item);
         if let Some(attr) = attr {
-            let item =
-                Annotatable::TraitItem(P(fully_configure!(self, item, noop_fold_trait_item)));
-            return self.collect_attr(attr, item, ExpansionKind::TraitItems).make_trait_items()
+            if !is_derive_attr(&attr) {
+                let item =
+                    Annotatable::TraitItem(P(fully_configure!(self, item, noop_fold_trait_item)));
+                return self.collect_attr(attr, item, ExpansionKind::TraitItems).make_trait_items()
+            } else {
+                self.cx.span_err(attr.span, "`derive` can't be only be applied to items");
```

Comment:
What would you return from`collect_attr` for the error case? The callers assume that `collect_attr` will always return a valid `Expansion`.